dashboard/modules/serve/serve_agent.py

```diff
@@ -74,6 +74,36 @@ async def get_all_deployments(self, req: Request) -> Response:
             content_type="application/json",
         )
 
+    @routes.get("/api/serve/applications/")
+    @optional_utils.init_ray_and_catch_exceptions()
+    async def get_serve_instance_details(self, req: Request) -> Response:
+        from ray.serve.schema import ServeInstanceDetails
+
+        controller = await self.get_serve_controller()
+
+        if controller is None:
+            # If no serve instance is running, return a dict that represents that.
+            details = ServeInstanceDetails.get_empty_schema_dict()
+        else:
+            try:
+                details = await controller.get_serve_instance_details.remote()
+            except ray.exceptions.RayTaskError as e:
+                # Task failure sometimes are due to GCS
+                # failure. When GCS failed, we expect a longer time
+                # to recover.
+                return Response(
+                    status=503,
+                    text=(
+                        "Fail to get the response from the controller. "
+                        f"Potentially the GCS is down: {e}"
+                    ),
+                )
+
+        return Response(
+            text=ServeInstanceDetails(**details).json(exclude_unset=True),
```

Comment:
Should this API return all the `ServeInstanceDetails` no matter the state? Is there a reason we're enabling `exclude_unset` here?

Reply:
Currently, my only concern is `route_prefix` at the deployment level as I mentioned [below](https://github.com/ray-project/ray/pull/33027#discussion_r1134289081), as this can be confusing for users. Since `get_serve_instance_details` (and the methods it calls) explicitly and intentionally fills in all info that we want to show to users (which is currently everything except the `route_prefix` at the deployment level), I chose to use `exclude_unset`.

Reply:
> Since `get_serve_instance_details` (and the methods it calls) explicitly and intentionally fills in all info that we want to show to users (which is currently everything except the `route_prefix` at the deployment level), I chose to use `exclude_unset`.

Makes sense– could you add this info to `get_serve_instance_details`'s docstring, and could you add a brief comment here explaining why you include `exclude_unset`?

By the way, is there a reason we need to convert `details` to a `ServeInstanceDetails` schema here? Couldn't we pass `details` directly into the `Response`?

Reply:
`details` is a Dict here, I think passing in the Dict and implicitly converting it to a string in Response doesn't work correctly. So I used `.json()`.

Reply:
Ah gotcha, could you change this to `json.dumps(details)` instead? That expresses the same logic more directly.

Reply:
Thanks! I also updated the docstring + added a note in the `get_serve_instance_details` method in the controller.